src/main/java/com/logsentinel/merkletree/utils/MathUtils.java

```diff
@@ -0,0 +1,42 @@
+package com.logsentinel.merkletree.utils;
+
+/**
+ * Math utilities
+ *
+ * @author Stefan Genchev
+ * @author https://stefan.genchev.io/
+ * @version 1.0
+ */
+public class MathUtils {
+    private static final double logSubtractFromNumber = 0.01;
+
+    /**
+     * Calculates the parameter k - the largest power of 2 smaller than n (i.e., k &lt; n &lt;= 2k)
+     *
+     * @param n the parameter n that is used to calculate the parameter k
+     * @return the calculated parameter k
+     */
+    public static int calculateK(int n) {
```

Comment:
Too many Math.log() is not very effective. Consider using something like:
Math.pow(2, (Integer.toBinaryString(n).length() -1 ))

Reply:
Optimization of the code sample is not necessary.